pkg/endpoint/api/endpoint_deletion_queue.go

```diff
@@ -0,0 +1,215 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/api/v1/server"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/endpointstate"
+	"github.com/cilium/cilium/pkg/lock/lockfile"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/promise"
+)
+
+type DeletionQueue struct {
+	logger                 *slog.Logger
+	lf                     *lockfile.Lockfile
+	endpointRestorePromise promise.Promise[endpointstate.Restorer]
+	endpointAPIManager     EndpointAPIManager
+	processed              chan struct{}
+}
+
+func (dq *DeletionQueue) Process(ctx context.Context, health cell.Health) error {
+	// hook context cancels when the start hooks have run, use context.Background()
+	// as we may be running after that.
+	if _, err := dq.endpointRestorePromise.Await(context.Background()); err != nil {
```

Comment:
I wonder whether processing the deletion-queue has to wait for the full restore of the endpoints (by calling `WaitForEndpointRestore` on the `Restorer`). Currently this isn't done - we just wait for the daemon promise (that we now replaced with waiting for the `Restorer`.